plugins/inputs/mdstat/mdstat.go

```diff
@@ -0,0 +1,304 @@
+// +build linux
+
+// Copyright 2018 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// Code has been changed since initial import.
+
+package mdstat
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"regexp"
+	"sort"
+	"strconv"
+	"strings"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/plugins/inputs"
+)
+
+const (
+	defaultHostProc = "/proc"
+	envProc         = "HOST_PROC"
+)
+
+var (
+	statusLineRE         = regexp.MustCompile(`(\d+) blocks .*\[(\d+)/(\d+)\] \[[U_]+\]`)
+	recoveryLineBlocksRE = regexp.MustCompile(`\((\d+)/\d+\)`)
+	recoveryLinePctRE    = regexp.MustCompile(`= (.+)%`)
+	recoveryLineFinishRE = regexp.MustCompile(`finish=(.+)min`)
+	recoveryLineSpeedRE  = regexp.MustCompile(`speed=(.+)[A-Z]`)
+	componentDeviceRE    = regexp.MustCompile(`(.*)\[\d+\]`)
+)
+
+type statusLine struct {
+	active int64
+	total  int64
+	size   int64
+	err    error
+}
+
+type recoveryLine struct {
+	syncedBlocks int64
+	pct          float64
+	finish       float64
+	speed        float64
+	err          error
+}
+
+type MdstatConf struct {
+	FileName string `toml:"file_name"`
+}
+
+func (k *MdstatConf) Description() string {
+	return "Get md array statistics from /proc/mdstat"
+}
+
+var mdSampleConfig = `
+	## Sets file path
+	## If not specified, then default is /proc/mdstat
+	# file_name = "/proc/mdstat"
+`
+
+func (k *MdstatConf) SampleConfig() string {
+	return mdSampleConfig
+}
+
+func evalStatusLine(deviceLine, statusLineStr string) statusLine {
+	sizeFields := strings.Fields(statusLineStr)
+	if len(sizeFields) < 1 {
+		return statusLine{active: 0, total: 0, size: 0,
+			err: fmt.Errorf("statusLine empty? %q", statusLineStr)}
+	}
+	sizeStr := sizeFields[0]
+	size, err := strconv.ParseInt(sizeStr, 10, 64)
+	if err != nil {
+		return statusLine{active: 0, total: 0, size: 0,
+			err: fmt.Errorf("unexpected statusLine %q: %w", statusLineStr, err)}
+	}
+
+	if strings.Contains(deviceLine, "raid0") || strings.Contains(deviceLine, "linear") {
```

Comment:
is it possible this match could be too liberal?

Reply:
I actually pulled that match line from the prometheus/procps library. If it _is_ too liberal, then it's broken there, too.

Reply:
And more specifically, that if statement is only looking at the "device" line. That line looks like this:
```
active raid1 sdb1[2] sda1[0]
```
So we should only see one array "type" on the line.